docs/ai/conceptual/understanding-tokens.md

```diff
@@ -0,0 +1,121 @@
+
+- Word tokenization (text is split into individual words based on a delimiter)
+- Character tokenization (text is split into individual characters)
+- Subword tokenization (text is split into partial words or character sets)
+
+For example, the GPT models, developed by OpenAI, use _Byte-Pair Encoding_ (BPE) which is a type of subword tokenization. OpenAI provides [a tool to visualize how text will be tokenized](https://platform.openai.com/tokenizer).
+
+There are benefits and disadvantages to each tokenization method:
+
+**Smaller tokens (character or subword tokenization)**
+
+- _Pros_:
+  - Allows the model to handle a wider range of inputs, such as unknown words, typos, or complex syntax
+  - Might allow the vocabulary size to be reduced, requiring less memory resources
+- _Cons_:
+  - A given text is broken into more tokens, requiring additional computational resources while processing
+  - Given a fixed token limit, the maximum size of the model's input and output is smaller
+
+**Larger tokens (word tokenization)**
+
+- _Pros_:
+  - A given text is broken into fewer tokens, requiring less computational resources while processing
+  - Given the same token limit, the maximum size of the model's input and output is larger
+- _Cons_:
+  - Might cause an increased vocabulary size, requiring more memory resources
+  - Can limit the models ability to handle unknown words, typos, or complex syntax
+
+## How tokens are used
+
+While training, after a text has been tokenized, each unique token is assigned an ID.
+
+Consider the following sentence:
+
+> I heard a dog bark loudly at a cat
+
+After tokenization, the following IDs could be assigned:
+
+- I (1)
+- heard (2)
+- a (3)
+- dog (4)
+- bark (5)
+- loudly (6)
+- at (7)
+- a (the "a" token is already assigned an ID of 3)
+- cat (8)
+
+By assigning IDs, text can be represented as a sequence of token IDs. The previous sentence would be represented as [1, 2, 3, 4, 5, 6, 7, 3, 8]. The sentence "I heard a cat" would be represented as [1, 2, 3, 8].
+
+As training continues, any new tokens in the training text are added to the vocabulary and assigned an ID. For example:
+
+- meow (9)
+- run (10)
+
+The semantic relationships between the tokens can be analyzed using these token ID sequences. These relationships are represented as multi-valued numeric vectors, known as [_embeddings_](understanding-embeddings.md). Each token is assigned an embedding based on how commonly it's used together with or in similar contexts to the other tokens.
+
+During output generation, the model predicts a vector value for the next token in the sequence. This vector value is then used to select the next token from the model's vocabulary. In practice, multiple vectors are calculated using various elements of the preceding tokens' embeddings. The model then evaluates all potential tokens from these vectors and selects the most probable one to continue the sequence.
```

Comment:
This article implies that an embedding is associated with an entire text, rather than a token. Which is correct? https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/understand-embeddings

Reply:
Each token has an embedding that represents its semantic attributes. Those semantic attributes are calculated by analyzing the entire text. But each embedding only represents the semantics of the associated token. This is the reference I pulled this info from: https://learn.microsoft.com/en-us/training/modules/fundamentals-generative-ai/3-language%20models#embeddings

Reply:
@alexwolfmsft just wanted to double check with you that my understanding is correct

Reply:
Hey @haywoodsloan - circling back to this. I'm not sure off hand if the embedding is associated with the entire text or a sum of all the individual token embeddings - I started looking into this but didn't find a confirmed answer. I will try to track down someone from the OpenAI service who might know for certain.

Reply:
Thanks for looking into this, I think there might be a difference between how embeddings are calculated when training a model vs when generating output. I can clarify that in this doc based on what you find.

Reply:
@elbruno - are you able to answer the question on whether the embedding is associated to the entire text or just a token?

Reply:
As I continue to research, I think I've found an answer. 

During training each token is given an embedding, these embeddings are 1-1 with the tokens (but they are calculated by analyzing the whole text). The important distinction being that each of those embeddings only represents an individual token.

After training, embeddings can be generated for larger blocks of text. The text is tokenized and the trained embeddings for each token are used to calculate an overall embedding score for the whole text. This embeddings represents the overall semantic meaning of the entire text. This is how we store and recall text memories, we match the embeddings of the question with the embedding of the memory.

So, in some sense both cases are true depending on the context. I'll add some details about how embeddings can represent larger texts once the model is trained.